### PR TITLE
Bug 1963079: Add e2e test for preferred host

### PR DESCRIPTION
**Pull request by damemi**
This adds a test for the patch in https://github.com/openshift/kubernetes/pull/638, copied from the same test implementation in KCM in https://github.com/openshift/cluster-kube-controller-manager-operator/pull/510/

### CODE REVIEW

**Review by damemi — COMMENTED** [review]
/cc @soltysh @p0lyn0mial 
ptal, I basically just copied Lukasz's KCM test and bumped some of our dependencies to get it working

File: test/e2e-preferred-host/ks_preferred_host_kas_test.go [hunk 1]
```diff
@@ -1,7 +1,188 @@
...
+	ksoOperatorConfig.Spec.UnsupportedConfigOverrides.Raw = raw
+	_, err = ksoOperator.Update(context.TODO(), ksoOperatorConfig, metav1.UpdateOptions{})
+	require.NoError(t, err)
+	testlibrary.WaitForPodsToStabilizeOnTheSameRevision(t, kubeClient.CoreV1().Pods(operatorclient.TargetNamespace), "kube-controller-manager=true", waitForKSORevisionSuccessThreshold, waitForKSORevisionSuccessInterval, waitForKSORevisionPollInterval, waitForKSORevisionTimeout)
+
+	// act
+	t.Log("creating a namespace and waiting for a default sa to be populated")
+	testNs, err := createTestingNS(t, "kcm-preferred-host", kubeClient)
+	require.NoError(t, err)
+	err = waitForServiceAccountInNamespace(kubeClient, testNs.Name, "default")
```
Comment (p0lyn0mial) [inline test/e2e-preferred-host/ks_preferred_host_kas_test.go:90]: I think we should create a `pod` instead. `KCM` deals with service accounts. Knowing that a pod has been scheduled gives us a signal that `KS` works over `localhost`. Does it make sense?

Comment (damemi) [reply test/e2e-preferred-host/ks_preferred_host_kas_test.go]: Yup, that makes sense. Updated (I also had a bunch of stale copypasta leftover from copying your test that I think I got now too)

File: test/e2e-preferred-host/ks_preferred_host_kas_test.go [hunk 2]
```diff
@@ -1,7 +1,188 @@
...
+			"unsupported-kube-api-over-localhost": []string{"true"},
+		},
+	}
+
+	raw, err := json.Marshal(data)
+	require.NoError(t, err)
+	ksoOperatorConfig.Spec.UnsupportedConfigOverrides.Raw = raw
+	_, err = ksoOperator.Update(context.TODO(), ksoOperatorConfig, metav1.UpdateOptions{})
+	require.NoError(t, err)
+	testlibrary.WaitForPodsToStabilizeOnTheSameRevision(t, kubeClient.CoreV1().Pods(operatorclient.TargetNamespace), "kube-scheduler=true", waitForKSORevisionSuccessThreshold, waitForKSORevisionSuccessInterval, waitForKSORevisionPollInterval, waitForKSORevisionTimeout)
```
Comment (p0lyn0mial) [inline test/e2e-preferred-host/ks_preferred_host_kas_test.go:84]: please change the pod selector to `scheduler=true`. We won't fine any pods that match `kube-scheduler=true`

File: test/e2e-preferred-host/ks_preferred_host_kas_test.go [hunk 3]
```diff
@@ -1,7 +1,188 @@
...
+
+	openShiftConfigClient, err := configv1client.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+
+	t.Log("getting the current Kube API host")
+	scheme, host := readCurrentKubeAPIHostAndScheme(t, openShiftConfigClient)
+
+	t.Logf("setting the \"unsupported-kube-api-over-localhost\" flag and pointing the current master to %q (non available host)", fmt.Sprintf("%s://%s:1234", scheme, host))
+	data := map[string]map[string][]string{
+		"extendedArguments": {
```
Comment (p0lyn0mial) [inline test/e2e-preferred-host/ks_preferred_host_kas_test.go:73]: you need to replace it with something like that:
```
unsupportedConfigOverrides:
  arguments:
    unsupported-kube-api-over-localhost:
      - "true"
    master:
      - "abc"
```


we added support in https://github.com/openshift/cluster-kube-scheduler-operator/pull/336/commits/e837452b29cd52b93748017b46bda4274a775260#r600414724

File: test/e2e-preferred-host/ks_preferred_host_kas_test.go [hunk 4]
```diff
@@ -1,7 +1,188 @@
...
+	ksoOperatorConfig.Spec.UnsupportedConfigOverrides.Raw = raw
+	_, err = ksoOperator.Update(context.TODO(), ksoOperatorConfig, metav1.UpdateOptions{})
+	require.NoError(t, err)
+	testlibrary.WaitForPodsToStabilizeOnTheSameRevision(t, kubeClient.CoreV1().Pods(operatorclient.TargetNamespace), "kube-scheduler=true", waitForKSORevisionSuccessThreshold, waitForKSORevisionSuccessInterval, waitForKSORevisionPollInterval, waitForKSORevisionTimeout)
+
+	// act
+	t.Log("creating a namespace and waiting for a default sa to be populated")
+	testNs, err := createTestingNS(t, "kso-preferred-host", kubeClient)
+	require.NoError(t, err)
+	err = waitForPodInNamespace(kubeClient, testNs.Name, "default")
```
Comment (p0lyn0mial) [inline test/e2e-preferred-host/ks_preferred_host_kas_test.go:90]: I think that there is no pod in the default namespace, you need to deploy one.

File: test/e2e-preferred-host/ks_preferred_host_kas_test.go [hunk 5]
```diff
@@ -1,7 +1,188 @@
...
+
+	raw, err := json.Marshal(data)
+	require.NoError(t, err)
+	ksoOperatorConfig.Spec.UnsupportedConfigOverrides.Raw = raw
+	_, err = ksoOperator.Update(context.TODO(), ksoOperatorConfig, metav1.UpdateOptions{})
+	require.NoError(t, err)
+	testlibrary.WaitForPodsToStabilizeOnTheSameRevision(t, kubeClient.CoreV1().Pods(operatorclient.TargetNamespace), "kube-scheduler=true", waitForKSORevisionSuccessThreshold, waitForKSORevisionSuccessInterval, waitForKSORevisionPollInterval, waitForKSORevisionTimeout)
+
+	// act
+	t.Log("creating a namespace and waiting for a default sa to be populated")
```
Comment (p0lyn0mial) [inline test/e2e-preferred-host/ks_preferred_host_kas_test.go:87]: please update the log msg.

File: test/e2e-preferred-host/ks_preferred_host_kas_test.go [hunk 6]
```diff
@@ -1,7 +1,188 @@
...
+	waitForKSORevisionSuccessThreshold = 3
+	waitForKSORevisionSuccessInterval  = 1 * time.Minute
+
+	// the following parameters specify max timeout after which
+	// KSOs are considered to not converged
+	waitForKSORevisionPollInterval = 30 * time.Second
+	waitForKSORevisionTimeout      = 7 * time.Minute
+
+	// podProvisionTimeout is how long to wait for a service account to be provisioned.
+	// service accounts are provisioned after namespace creation
```
Comment (p0lyn0mial) [inline test/e2e-preferred-host/ks_preferred_host_kas_test.go:46]: please update the comment.

File: test/e2e-preferred-host/ks_preferred_host_kas_test.go [hunk 7]
```diff
@@ -1,7 +1,188 @@
...
+
+var (
+	// the following parameters specify for how long apis must
+	// stay on the same revision to be considered stable
+	waitForKSORevisionSuccessThreshold = 3
+	waitForKSORevisionSuccessInterval  = 1 * time.Minute
+
+	// the following parameters specify max timeout after which
+	// KSOs are considered to not converged
+	waitForKSORevisionPollInterval = 30 * time.Second
```
Comment (p0lyn0mial) [inline test/e2e-preferred-host/ks_preferred_host_kas_test.go:42]: this is more like `KS`, we will be waiting for the scheduler pods, not the operator.

**Comment by damemi** [comment]
Updated with feedback, @p0lyn0mial thanks is there anything else I missed?

**Comment by p0lyn0mial** [comment]
/lgtm

thanks for the PR!

**Pull request merged**